src/Features/CSharp/Portable/ExtractMethod/Extensions.cs

```diff
@@ -85,18 +94,48 @@ bool predicate(SyntaxNode n)
 
                 return false;
             }
+        }
 
-            if (!node.GetAncestorsOrThis<SyntaxNode>().Any(predicate))
+        public static bool ContainedInValidType(this SyntaxNode node)
+        {
+            Contract.ThrowIfNull(node);
+
+            if (IsContainedDirectlyInNamespace(node))
             {
                 return false;
             }
 
-            if (node.FromScript() || node.GetAncestor<TypeDeclarationSyntax>() != null)
+            return true;
+
+            static bool IsContainedDirectlyInNamespace(SyntaxNode n)
             {
-                return true;
-            }
+                var hasContainer = false;
+                var immediateParent = false;
+                while (n is not null)
+                {
+                    if (n is NamespaceDeclarationSyntax && hasContainer && immediateParent)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        immediateParent = false;
+                    }
+
+                    if (n is BaseMethodDeclarationSyntax
+                       or AccessorDeclarationSyntax
+                       or BlockSyntax
+                       or CompilationUnitSyntax)
+                    {
+                        hasContainer = true;
+                        immediateParent = true;
+                    }
+
+                    n = n.Parent!;
```

Comment:
i do not know what is going on here.  in english what are you trying to express :)

Reply:
if I am a method, or a block, etc. and my immediate parent is a namespace, then bail and do not allow that as a valid selection. Still need to make sure I am handling everything that shouldn't be directly in a namespace

Reply:
would it suffice to just ask: did i hit a namespace, but i did not hit a type?  that seems simpler.  or, put another way:

first look if we're contained by a type.  if so, great.
then look to see if we're contained by a namespace.  if so, bad.
otherwise, we must be global.  if so, great.